lib/rouge/lexers/bbcbasic.rb

```diff
@@ -0,0 +1,115 @@
+# -*- coding: utf-8 -*- #
+# frozen_string_literal: true
+
+module Rouge
+  module Lexers
+    class BBCBASIC < RegexLexer
+      title "BBCBASIC"
+      desc "BBC BASIC syntax"
+      tag 'bbcbasic'
+      filenames '*,fd1'
+
+      def self.punctuation
+        @punctuation ||= %w(
+          [:,;'~] SPC TAB
+        )
+      end
+
+      def self.function
+        @function ||= %w(
+          ABS ACS ADVAL ASC ASN ATN BEATS BEAT BGET# CHR\$ COS COUNT DEG EOF#
+          ERL ERR ERROR EVAL EXP EXT# GET GET\$ GET\$# HIMEM INKEY INKEY\$
+          INSTR INT LEFT\$ LEN LN LOG LOMEM MID\$ OPENIN OPENOUT OPENUP PAGE
+          POS PTR# RAD REPORT\$ RIGHT\$ RND SGN SIN SQR STR\$ STRING\$ SUM
+          SUMLEN TAN TEMPO TIME TIME\$ TOP USR VAL VPOS
+        )
+      end
+
+      def self.control
+        @control ||= %w(
+          CASE CHAIN ELSE ENDCASE ENDIF ENDPROC ENDWHILE END FN FOR GOSUB GOTO
+          IF INSTALL LIBRARY NEXT OF OTHERWISE OVERLAY PROC REPEAT RETURN STEP
+          STOP THEN TO UNTIL WHEN WHILE
+        )
+      end
+
+      def self.statement
+        @statement ||= %w(
+          BEATS BPUT# CALL CLEAR CLG CLOSE# CLS COLOR COLOUR DATA DIM ENVELOPE
+          GCOL LET MODE OFF ON ORIGIN OSCI PLOT PRINT PRINT# QUIT READ REPORT
+          SOUND STEREO SWAP SYS TINT VDU VOICE VOICES WAIT WIDTH
+        )
+      end
+
+      def self.operator
+        @operator ||= %w(
+          << <= <> < >= >>> >> > [-!\$()*+\/=?^|] AND DIV EOR MOD NOT OR
+        )
+      end
+
+      def self.constant
+        @constant ||= %w(
+          FALSE TRUE
+        )
+      end
+
+      state :root do
+        rule %r/[ \n]+/, Text
+        rule %r/[\[]/, Keyword, :assembly1
+        rule %r/\*.*/, Generic::Prompt # CLI command
+        rule %r/REM *>.*/, Comment::Special
+        rule %r/REM.*/, Comment
+        rule %r/#{BBCBASIC.punctuation.join('|')}/, Punctuation
+        rule %r/#{BBCBASIC.function.join('|')}/, Name::Builtin # function or pseudo-variable
+        rule %r/(?:DIM|POINT)(?=\()/, Name::Builtin # function sharing keyword with statement, distinguished by ()
```

Comment:
Would `\b` be a better way to do it?

Reply:
I'm trying to distinguish between `DIM` and `POINT` keywords when used in their imperative form, and when they act as a function, returning a value. The only condition under which they act as functions is if the immediately following character is a `(`. The following examples are all syntactically valid:

String              | Meaning | Would match `DIM\b`?
---------------|---------|-------------------------
`DIMx 100`      | statement | no
`DIM y(100)`    | statement | yes
``DIM`z 100`` | statement | yes
`DIM(y())`         | function    | yes
`DIM(y(),0)`     | function | yes

So I think the answer is no, `\b` can't be used instead.

Reply:
Makes sense.